forem/forem
# Expand @-mention functionality to posts

## What type of PR is this? (check all applicable)

## Description

This PR completes the final piece of functionality to _fully_ ship [RFC #22](https://github.com/forem/rfcs/pull/22)! 🥰 

Per the RFC's [design/definition of done](https://github.com/forem/rfcs/blob/main/text/0022-expand-at-mention-to-posts.md#designimplementation), this PR does the following:
- [x] Allows articles to have mentions
- [x] Sends users who are mentioned in an article a single notification
- [x] Ensures that the correct notification shows up on the mentioned user's `/notification` page
- [x] Ensures that a user only receives one notification per article (_if a user follows an author/organization but also happen to be @-mentioned in the article, they _only_ receive the @-mention notification, and no other article notifications_)
- [x] Updates relevant views and mailer templates
- [x] Adds a TON of specs

## Related Tickets & Documents

Completes https://github.com/forem/rfcs/pull/22! 🎉 

## QA Instructions, Screenshots, Recordings

**To QA this PR, open up two browser windows, each with two different users. In this case, I have User A (Admin user) and User B (my user, `vaidehijoshi`).**

### Part 1: Testing @-mentions while following a user
1. Start by having User B follow User A. In User A's account, create an article, and mention User B:
<img width="812" alt="Screen Shot 2021-05-10 at 11 07 32 AM" src="https://user-images.githubusercontent.com/6921610/117854421-5349e400-b23e-11eb-9763-eacb04ed1f5c.png">
2. Publish the article written by User A:
<img width="1138" alt="Screen Shot 2021-05-10 at 11 07 47 AM" src="https://user-images.githubusercontent.com/6921610/117854628-85f3dc80-b23e-11eb-9259-46dd12a9b9ee.png">
3. Go to User B's account, and check that you see a notification icon:
<img width="288" alt="Screen Shot 2021-05-10 at 11 14 36 AM" src="https://user-images.githubusercontent.com/6921610/117854644-89876380-b23e-11eb-8347-6e54a0a62bf4.png">
4. Click on the icon, which takes you to the `/notifications` tab. You should see a notification that tells you that you were mentioned. You SHOULD NOT have received a notification telling you that User A published the article (which you normally would get if you follow User A). Instead, you should _only_ have a notification telling you that you were mentioned in the new post they published:
<img width="757" alt="Screen Shot 2021-05-10 at 11 08 11 AM" src="https://user-images.githubusercontent.com/6921610/117854769-a91e8c00-b23e-11eb-92ae-b53acb545960.png">
5. Clicking on the notification itself should take you to the newly-published article, and you should see your username mentioned 👍🏽 
<img width="1074" alt="Screen Shot 2021-05-10 at 11 08 27 AM" src="https://user-images.githubusercontent.com/6921610/117854791-ad4aa980-b23e-11eb-8cae-98718c647027.png">

### Part 2: Testing normal article published notifications
1. Now, create a new article from User A's account. This time, DO NOT mention User B.
<img width="1123" alt="Screen Shot 2021-05-10 at 11 15 51 AM" src="https://user-images.githubusercontent.com/6921610/117855195-19c5a880-b23f-11eb-9d59-c5854296b848.png">
2. In User B's account, check that you see a notification icon:
<img width="288" alt="Screen Shot 2021-05-10 at 11 14 36 AM" src="https://user-images.githubusercontent.com/6921610/117855301-32ce5980-b23f-11eb-90cb-4a6e526123d5.png">
3. You should see a normal "article published" notification:
<img width="1030" alt="Screen Shot 2021-05-10 at 11 14 45 AM" src="https://user-images.githubusercontent.com/6921610/117855210-1cc09900-b23f-11eb-96a8-ff43299810cb.png">

### Part 3: Testing embedded mentions
1. Find a comment that mentions User B (any comment should do!). For example:
<img width="1266" alt="Screen Shot 2021-05-10 at 1 59 12 PM" src="https://user-images.githubusercontent.com/6921610/117855486-61e4cb00-b23f-11eb-9682-a12f7ca0bfb5.png">
2. Create a new post from User A's account, using the `{% comment [id] %}` liquid tag syntax:
<img width="814" alt="Screen Shot 2021-05-10 at 1 59 44 PM" src="https://user-images.githubusercontent.com/6921610/117855775-ab351a80-b23f-11eb-86af-a407a67ca6c4.png">
3. Go to User B's account; if User B is still following User A, there should be a new notification, but it should NOT be a notification about being @-mentioned. Instead, it should only be a normal "article published" notification, and nothing else. 
<img width="752" alt="Screen Shot 2021-05-10 at 2 00 02 PM" src="https://user-images.githubusercontent.com/6921610/117855788-aff9ce80-b23f-11eb-950a-2d207879f78e.png">
4. You can try the same test after making User B "unfollow" User A, at which point User B should receive NO notifications.

### Part 4: Other oddities
1. Try testing this functionality out with both editors; the functionality should be exactly the same:
<img width="1411" alt="Screen Shot 2021-05-10 at 11 26 40 AM" src="https://user-images.githubusercontent.com/6921610/117856064-0535e000-b240-11eb-844d-54c73aa36504.png">
<img width="1036" alt="Screen Shot 2021-05-10 at 12 59 35 PM" src="https://user-images.githubusercontent.com/6921610/117856052-023aef80-b240-11eb-834a-32f81ad8692f.png">
2. Test that a user doesn't get more than one notification, no matter how many times they are mentioned! ALso test that you can't @-mention more than 7 times:
<img width="1266" alt="Screen Shot 2021-05-10 at 11 24 05 AM" src="https://user-images.githubusercontent.com/6921610/117856155-1f6fbe00-b240-11eb-878e-9b475c6844a3.png">
3. Test that you can [preview the new `new_article_mention_email` mailer](http://localhost:3000/rails/mailers/notify_mailer) locally, and that the job to send the email is enqueued when a mention notification is sent locally.
<img width="967" alt="Screen Shot 2021-05-10 at 11 18 25 AM" src="https://user-images.githubusercontent.com/6921610/117856252-3ca48c80-b240-11eb-96e7-d6c202705a8c.png">
<img width="1423" alt="Screen Shot 2021-05-10 at 11 31 38 AM" src="https://user-images.githubusercontent.com/6921610/117856371-652c8680-b240-11eb-961d-5a3bfad47760.png">


### UI accessibility concerns?

_This is mostly backend work, and the frontend work is reusing HTML markup code that already exists so, hopefully not!_

## Added tests?

- [x] You bet I did 😉 

## [Forem core team only] How will this change be communicated?
      or in a [forem.dev](http://forem.dev) post

## Are there any post deployment tasks we need to perform?

Nothing special, nope! Just want to be sure to QA thoroughly and try it out on DEV before rolling it out to the fleet.

## What gif best describes this PR or how it makes you feel?

![omg sloth](https://media2.giphy.com/media/lRpD28vbtrqq2zpaKS/giphy.gif?cid=790b7611d9cb3f7b2095ed55fa1b94ec0f7c925ce7b1dafd&rid=giphy.gif&ct=g)
